File: libs/iqrf/cdc-iqrf/java/jlibcdc-examples/src/main/java/com/microrisc/cdc/examples/WriteRead.java
/* 
 * Copyright 2014 MICRORISC s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.microrisc.cdc.examples;

import com.microrisc.cdc.*;
import java.util.Arrays;

/**
 * Example demonstrates sending requests for actual temperature on node's sensor 
 * and asynchronous reception of actual temperature's values.  
 */
public class WriteRead implements J_AsyncMsgListener {
    
    @Override
    public void onGetMessage(short[] data) {
        // positions of fields
        final int INT_VALUE_POS = 8;
        final int FULL_VALUE_POS = 9;
        final int FULL_VALUE_LENGTH = 2;
        
        // for DPA request read temperature
        final short[] CONFIRMATION_HEADER = { 0x01, 0x00, 0x0A, 0x00 };
        final short[] RESPONSE_HEADER = { 0x01, 0x00, 0x0A, 0x80 };
        
        // no data
        if (data == null || data.length == 0) {
            System.out.println("No data received\n");
            return;
        }
        
        // data length
        System.out.println("data length: " + data.length);
        
        // response flag
        boolean responseReceived = false;
        
        // header of data
        short[] dataHeader = Arrays.copyOfRange(data, 0, 4);
        
        if (Arrays.equals(dataHeader, CONFIRMATION_HEADER)) {
            System.out.print("confirmation received: ");
        } else if (Arrays.equals(dataHeader, RESPONSE_HEADER)) {
            System.out.print("response received: ");
            responseReceived = true;
        } else {
            System.out.print("unknown type of message received: ");
        }
        
        // display raw data
        for (short sh : data) {
            System.out.print(Integer.toHexString(sh) + " ");
        }
        System.out.println();
        
        // parse data to get temperature
        if (responseReceived) {
            short value = data[INT_VALUE_POS];
            
            short[] fullTemperatureValue = new short[FULL_VALUE_LENGTH];
            System.arraycopy(data, FULL_VALUE_POS, fullTemperatureValue, 0, FULL_VALUE_LENGTH);
            byte fractialPart = (byte)(fullTemperatureValue[0] & 0x0F);
        
            System.out.println("Temperature = " + value + "." + fractialPart + " C");
        }
        
        System.out.println();
    }
    
    
    static public void main(String[] args) {
        J_CDCImpl myCDC = null;
        try {
            // creating CDC object, which will communicate via /dev/ttyACM0
            //myCDC = new J_CDCImpl("/dev/ttyACM0");
            myCDC = new J_CDCImpl("COM4");
            
            // communication testing
            if (myCDC.test()) {
                System.out.println("Test OK");
            } else {
                System.out.println("Test FAILED");
                myCDC.destroy();
                return;
            }
        } catch (Exception ex) {
            System.out.println(ex.toString());
            if (myCDC != null) {
                myCDC.destroy();
            }
            return;
        }
        
        // listener of asynchronous messages
        WriteRead example = new WriteRead();
        
        // register to receiving asynchronous messages
        myCDC.registerAsyncListener(example);
            
        // Send DPA read temperature request
        // NAdr=0x01 0x00 PNum=0x0A PCmd=0x00
        short[] temperatureRequest = { 0x01, 0x00, 0x0A, 0x00, 0xFF, 0xFF };
        
        for (int sendCounter = 0; sendCounter < 10; sendCounter++) {
            try {
                // sending read temperature request and checking response of the device
                J_DSResponse response = myCDC.sendData(temperatureRequest);
                if (response != J_DSResponse.OK) {
                    // processing of bad response ...
                    System.out.println("Response not OK: " + response);
                }
            } catch (J_CDCSendException ex) {
                System.out.println("Send error occurred: " + ex.toString());
                // send exception processing...
            } catch (J_CDCReceiveException ex) {
                System.out.println("Receive error occurred: " + ex.toString());
                // receive exception processing...
            } catch (Exception ex) {
                System.out.println("Other error occurred: " + ex.toString());
                // other exception processing...
            }
            
            // if reception is stopped, is not further possible to send and 
            // to receive any next messages
            if (myCDC.isReceptionStopped()) {
                myCDC.destroy();
                return;
            }
            
            // sending another message and waiting for the answer
            // we should wait for a while - to give USB device time to get ready 
            // to succesfully process next message
            try {
                Thread.sleep(1000);
            } catch (InterruptedException ex) {
                System.out.println("Thread interrupted\n");
                break;
            }
        }
        
        // terminate library and free up used resources
        myCDC.destroy();
    }
}

File: libs/iqrf/cdc-iqrf/c++/win/libcdc-example/WriteRead.cpp
/*
 * Copyright 2015 MICRORISC s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * DPA Write-Read example
 *
 * @author      Michal Konopa, Rostislav Spinar
 * @version     1.0.0
 * @date        1.4.2015
 */

#include <iostream>
#include <windows.h>
#include <cdc/CDCImpl.h>


// length of message header
const unsigned int HEADER_LENGTH = 4;

// confirmation header
const unsigned char CONFIRMATION_HEADER[] = { 0x01, 0x00, 0x0A, 0x00 };

// response header
const unsigned char RESPONSE_HEADER[] = { 0x01, 0x00, 0x0A, 0x80 };

// prints specified data onto standard output in hex format
void printDataInHex(unsigned char* data, unsigned int length) {
    for ( int i = 0; i < length; i++ ) {
        std::cout << "0x" << std::hex << (int)*data;
        data++;
        if ( i != (length - 1) ) {
            std::cout << " ";
        }
    }
    std::cout << std::dec << "\n";
}

// determines, if the specified message header is the CONFIRMATION
bool isConfirmation(unsigned char* msgHeader) {
    for ( int i = 0; i < HEADER_LENGTH; i++ ) {
        if ( *msgHeader != CONFIRMATION_HEADER[i] ) {
            return false;
        }
        msgHeader++;
    }
    return true;
}

// determines, if the specified message header is the RESPONSE
bool isResponse(unsigned char* msgHeader) {
    for ( int i = 0; i < HEADER_LENGTH; i++ ) {
        if ( *msgHeader != RESPONSE_HEADER[i] ) {
            return false;
        }
        msgHeader++;
    }
    return true;
}

// prints specified response onto standard output
void printResponse(unsigned char* response, unsigned int length) {
    // positions of fields in the response
    const int ERROR_CODE_POS = 6;
    const int ERROR_CODE_OK = 0;
    const int INT_VALUE_POS = 8;
    const int FULL_VALUE_POS = 9;
    const int FULL_VALUE_LENGTH = 2;

    // checking of ErrN field
    int errorCode = *(response+ERROR_CODE_POS);
    if ( errorCode != ERROR_CODE_OK ) {
        std::cout << "Error: " << errorCode << "\n";
        return;
    }

    unsigned char value = *(response + INT_VALUE_POS);

    unsigned char* fullTemperatureValue = new unsigned char[FULL_VALUE_LENGTH];
    memcpy(fullTemperatureValue, response+FULL_VALUE_POS, FULL_VALUE_LENGTH);
    unsigned char fractialPart = *fullTemperatureValue;

    std::cout << "Temperature = " << (int)value << "." << (int)fractialPart << " C" << "\n";
    delete[] fullTemperatureValue;
}

void receiveData(unsigned char* data, unsigned int length) {
    if ( data == NULL || length == 0 ) {
        std::cout << "No data received\n";
        return;
    }

    if ( length < HEADER_LENGTH ) {
        std::cout << "Unknown data: ";
        printDataInHex(data, length);
        return;
    }

    unsigned char* msgHeader = new unsigned char[HEADER_LENGTH];
    memcpy(msgHeader, data, HEADER_LENGTH);

    if ( isConfirmation(msgHeader) ) {
        std::cout << "Confirmation received: ";
        printDataInHex(data, length);
        delete[] msgHeader;
        return;
    }

    if ( isResponse(msgHeader) ) {
        std::cout << "Response received: \n";
        printResponse(data, length);
        delete[] msgHeader;
        return;
    }

	std::cout << "Unknown type of message. Data: ";
	printDataInHex(data, length);
	delete[] msgHeader;
}


int main() {
	CDCImpl* testImp = NULL;
	try {
		testImp = new CDCImpl("COM5");

		bool test = testImp->test();

		if ( test ) {
			std::cout << "Test OK\n";
		} else {
			std::cout << "Test FAILED\n";
			delete testImp;
			return 2;
		}
	} catch ( CDCImplException& e ) {
		std::cout << e.getDescr() << "\n";
		if ( testImp != NULL ) {
			delete testImp;
		}
		return 1;
	}

	// register to receiving asynchronous messages
	testImp->registerAsyncMsgListener(&receiveData);

	// data to send to USB device
	const int REQUEST_LENGTH = 6;
	unsigned char temperatureRequest[REQUEST_LENGTH] = { 0x01, 0x00, 0x0A, 0x00, 0xFF, 0xFF };

	for ( int sendCounter = 0; sendCounter < 10; sendCounter++ ) {
		try {
			// sending read temperature request and checking response of the device
			DSResponse dsResponse = testImp->sendData(temperatureRequest, REQUEST_LENGTH);
			if ( dsResponse != OK ) {
				// bad response processing...
				std::cout << "Response not OK: " << dsResponse << "\n";
			}
		} catch ( CDCSendException& ex ) {
			std::cout << ex.getDescr() << "\n";
			// send exception processing...
		} catch ( CDCReceiveException& ex ) {
			std::cout << ex.getDescr() << "\n";
			// receive exception processing...
		}

		// if reception is stopped, is not further possible to send and
		// to receive any next messages
        if ( testImp->isReceptionStopped() ) {
            delete testImp;
            return 1;
        }

        // sending another message and waiting for the answer
        // we should wait for a while - to give USB device time to get ready
        // to succesfully process next message
        Sleep(1000);
    }

    delete testImp;
    return 0;
}


